MCMC: export samples to ASCII file

dbkeitel opened the pull request:
As per #7 this adds simple ASCII export of MCMC samples. Passes the tests and looks sane in the first few example cases, but I will check how it works on a real-life follow-up example before merging.

Uses the same file header as GridSearch.

dbkeitel commented:
Now also works for MCMCTransientSearch and the main use case, MCMCFollowUpSearch.

dbkeitel commented on pyfstat/core.py:
```diff
@@ -224,9 +224,12 @@ def predict_fstat(
     """ The base search class providing parent methods to other searches """
 
-    def _add_log_file(self):
+    def _add_log_file(self, header=[]):
```
Thanks, I wasn't aware of that weird scoping issue! It seems here there's no problem, since the header list is not actually modified inside the function. (I think you're referring to the "if the function modifies the object (e.g. by appending an item to a list), the default value is in effect modified" point in the docs, right?) But it might still be good to implement your suggestion, and I can later audit the rest of the code for possible harmful instances of this type of issue.

Rodrigo-Tenorio replied:
Yes, that's what I mean. It shouldn't be an issue in this context, but I think is worth fixing, given that no code relies on this yet.

dbkeitel replied:
Can you put it here as a code suggestion? I'd like to see how that feature over here compares to the one from gitlab.

(Let me know if you need to be added as a project member to do so.)

Rodrigo-Tenorio replied:
I've tried to add it myself, but it seems I have to be a member

dbkeitel replied:
I've asked Greg to add you, but have committed the change directly in the meantime. Thanks!

Rodrigo-Tenorio commented:
Codes run as expected. Output results contain proper header and VC information.

- Prevent the examples from ending with an error in case latex was not found. LaTeX will be used by matplotlib if present. 
- Moreover, corner plot requires matplotlib < 3.0 due to deprecated options being used.

@dbkeitel , the plot supposed to show the walkers is empty. I guess it has to do with `_plot_walker()` at pyfstat/mcmc_based_searches.py:1196. I've checked the follow up example
without modifications and the produced plot is empty as well. Do you know something about this?

dbkeitel commented:
@Rodrigo-Tenorio sorry for the delay.

>    Added binary parameters to the built-in ones (although one can make it's own parameters through inheritance, I think it makes sense to have them this way, as even makefakedata includes them).

Looks ok.

>    Prevent the examples from ending with an error in case latex was not found. LaTeX will be used by matplotlib if present.

Thanks, that's useful!

>    Moreover, corner plot requires matplotlib < 3.0 due to deprecated options being used.

I'll revert that change from this branch for the moment, because I'm not seeing problems with corner plots on matplotlib 3.1.3. But please post further details on #31.

> the plot supposed to show the walkers is empty. I guess it has to do with _plot_walker() at pyfstat/mcmc_based_searches.py:1196. I've checked the follow up example
without modifications and the produced plot is empty as well. Do you know something about this?

It was working for me with an earlier version of this branch, and is still working with other examples e.g. `examples/mcmc_examples/fully_coherent_search_using_MCMC.py` but with `examples/followup_examples/semi_coherent_directed_follow_up.py` it's empty as you report. I'll investigate.

> Also, I see that binary parameters are not supported for injection using this code... Since injections using noise SFT seem not to be supported (am I right?), there is need to focus on this issue by now.

Can you please open separate issues/PRs for any parts of the code where explicit binary support needs to be added, after this one is merged?

dbkeitel commented:
I've squashed your commits into 2 and removed the matplotlib version change.

>> the plot supposed to show the walkers is empty. I guess it has to do with _plot_walker() at pyfstat/mcmc_based_searches.py:1196. I've checked the follow up example
    without modifications and the produced plot is empty as well. Do you know something about this?

> It was working for me with an earlier version of this branch, and is still working with other examples e.g. examples/mcmc_examples/fully_coherent_search_using_MCMC.py but with examples/followup_examples/semi_coherent_directed_follow_up.py it's empty as you report. I'll investigate.

The walker plot actually works when running the example the first time, but becomes empty when rerun on saved data, as it doesn't have the chain history available then. I'll put in an error or skip check.

dbkeitel commented:
> The walker plot actually works when running the example the first time, but becomes empty when rerun on saved data, as it doesn't have the chain history available then. I'll put in an error or skip check.

The problem is with the `return_fig` option that only `MCMCFollowUpSearch.run()` has, not `MCMCSearch.run()`, and then explicitly writing out the plot file in the example script regardless of whether the plot has bee updated.

The simple fix would be to change the example script to use the default internal plotting; but I'll rather merge this now as is and properly fix the return state of the plotting in a separate PR.
